### ISSUE TITLE
overlay network sometimes not work between docker service containers

### ISSUE THREAD

**Issue by hehailong5** [moby/moby#30304]
version: 1.12.3

I've created three docker services, all attached to a overlay network. it is observed that sometimes the network is unreachable from each other.

1) service A (can reach service B, can not reach service C):

```
/opt/application # ip add
13: eth0@if14: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1450 qdisc noqueue state UP 
    link/ether 02:42:0a:00:00:03 brd ff:ff:ff:ff:ff:ff
    inet 10.0.0.3/24 scope global eth0
       valid_lft forever preferred_lft forever
    inet 10.0.0.2/32 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:aff:fe00:3/64 scope link 
       valid_lft forever preferred_lft forever
/opt/application # 
/opt/application # ping 10.0.0.5
PING 10.0.0.5 (10.0.0.5): 56 data bytes
64 bytes from 10.0.0.5: seq=0 ttl=64 time=1.345 ms
64 bytes from 10.0.0.5: seq=1 ttl=64 time=0.301 ms
^C
--- 10.0.0.5 ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.301/0.823/1.345 ms
/opt/application # 
/opt/application # ping 10.0.0.7
PING 10.0.0.7 (10.0.0.7): 56 data bytes
^C
```

2) service B (can reach service A, can not reach service C):

```
/opt/application # ip add
13: eth0@if14: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1450 qdisc noqueue state UP 
    link/ether 02:42:0a:00:00:05 brd ff:ff:ff:ff:ff:ff
    inet 10.0.0.5/24 scope global eth0
       valid_lft forever preferred_lft forever
    inet 10.0.0.4/32 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:aff:fe00:5/64 scope link 
       valid_lft forever preferred_lft forever
/opt/application # ping 10.0.0.3
PING 10.0.0.3 (10.0.0.3): 56 data bytes
64 bytes from 10.0.0.3: seq=0 ttl=64 time=1.512 ms
64 bytes from 10.0.0.3: seq=1 ttl=64 time=0.339 ms
^C
/opt/application # ping 10.0.0.7
PING 10.0.0.7 (10.0.0.7): 56 data bytes
^C
```

3) service C (can not reach service A and service B):

```
/opt/application # ip add
13: eth0@if14: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1450 qdisc noqueue state UP 
    link/ether 02:42:0a:00:00:07 brd ff:ff:ff:ff:ff:ff
    inet 10.0.0.7/24 scope global eth0
       valid_lft forever preferred_lft forever
    inet 10.0.0.6/32 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:aff:fe00:7/64 scope link 
       valid_lft forever preferred_lft forever
/opt/application # ping 10.0.0.3
PING 10.0.0.3 (10.0.0.3): 56 data bytes
^C
--- 10.0.0.3 ping statistics ---
8 packets transmitted, 0 packets received, 100% packet loss
/opt/application # 
/opt/application # ping 10.0.0.5
PING 10.0.0.5 (10.0.0.5): 56 data bytes
^C
```

**Comment by gianarb** [comment]
Hello, thanks for your report. Can we have more information about the topology of your cluster?
How many nodes are you using? Can you paste `docker info` from your master?

Thanks

**Comment by hehailong5** [comment]
Hi,

this is a 3 nodes cluster.

here goes the "docker info":

root@paas-controller:~# docker info
Containers: 50
 Running: 37
 Paused: 0
 Stopped: 13
Images: 28
Server Version: 1.12.3
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 355
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null bridge host overlay
Swarm: active
 NodeID: erpmpgxyk9go1c9drvb4obsi6
 Is Manager: true
 ClusterID: 8kbklo1ke0iov6s2320k40lm6
 Managers: 3
 Nodes: 3
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 120.0.0.206
Runtimes: runc
Default Runtime: runc
Security Options: apparmor
Kernel Version: 4.4.0-31-generic
Operating System: Ubuntu 14.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 7.796 GiB
Name: paas-controller
ID: WIWO:PYNI:O2L7:UPA3:UIF5:5P5S:IALU:S5CI:R6XM:ZKDM:CYAE:B3P7
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 110.0.0.21:6666
 127.0.0.0/8
root@paas-controller:~#

**Comment by thaJeztah** [comment]
Are you only having this issue when using `ping`? Pinging services across nodes may not work, due to the way Virtual IP's are implemented; are you having the same issue if you (e.g.) `curl` the other service?

I also recommend upgrading your daemons to the latest 1.12 version (1.12.6). Many bug fixes have been made since 1.12.3, and 1.12.6 fixes a CVE.

**Comment by hehailong5** [comment]
Actually the nc command failed in my startup script.
OK, I will try 1.12.6

**Comment by hehailong5** [comment]
the same happens today. I am using 1.13.1 this time.

**Comment by thaJeztah** [comment]
Let try to get some more information to work with;

- When you initialised the swarm cluster, did you use the `--advertise-addr` option?
- Did you use any options when creating the overlay network that the services are attached to? (e.g. `--o encrypted`
- How are the tasks distributed for those services; e.g. are the tasks for service C on a specific node?
- Are all required ports open between the nodes (https://docs.docker.com/engine/swarm/swarm-tutorial/#open-protocols-and-ports-between-the-hosts)?
- Where are the nodes running? e.g. bare metal, in VM's, on AWS, Azure, DigitalOcean?

Can you try running the `check-config.sh` script on the nodes to see if anything stands out? https://raw.githubusercontent.com/docker/docker/master/contrib/check-config.sh

**Comment by hehailong5** [comment]
Hi,

- When you initialized the swarm cluster, did you use the --advertise-addr option?

YES.

- Did you use any options when creating the overlay network that the services are attached to? (e.g. --o encrypted

NO.

- How are the tasks distributed for those services; e.g. are the tasks for service C on a specific node?

I have 3 services, each was running in replicated  1/1 mode, AND, each was attached to a specific node using labels.


- Are all required ports open between the nodes (https://docs.docker.com/engine/swarm/swarm-tutorial/#open-protocols-and-ports-between-the-hosts)?

YES.

- Where are the nodes running? e.g. bare metal, in VM's, on AWS, Azure, DigitalOcean?

VM

- Can you try running the check-config.sh script on the nodes to see if anything stands out? https://raw.githubusercontent.com/docker/docker/master/contrib/check-config.sh


here is the output:
-------------------------------------------------------------------------------------------
-bash-4.2# ./check.sh 
warning: /proc/config.gz does not exist, searching other paths for kernel config ...
info: reading kernel config from /boot/config-3.10.0-514.6.1.el7.x86_64 ...

Generally Necessary:
- cgroup hierarchy: properly mounted [/sys/fs/cgroup]
- CONFIG_NAMESPACES: enabled
- CONFIG_NET_NS: enabled
- CONFIG_PID_NS: enabled
- CONFIG_IPC_NS: enabled
- CONFIG_UTS_NS: enabled
- CONFIG_CGROUPS: enabled
- CONFIG_CGROUP_CPUACCT: enabled
- CONFIG_CGROUP_DEVICE: enabled
- CONFIG_CGROUP_FREEZER: enabled
- CONFIG_CGROUP_SCHED: enabled
- CONFIG_CPUSETS: enabled
- CONFIG_MEMCG: enabled
- CONFIG_KEYS: enabled
- CONFIG_VETH: enabled (as module)
- CONFIG_BRIDGE: enabled (as module)
- CONFIG_BRIDGE_NETFILTER: enabled (as module)
- CONFIG_NF_NAT_IPV4: enabled (as module)
- CONFIG_IP_NF_FILTER: enabled (as module)
- CONFIG_IP_NF_TARGET_MASQUERADE: enabled (as module)
- CONFIG_NETFILTER_XT_MATCH_ADDRTYPE: enabled (as module)
- CONFIG_NETFILTER_XT_MATCH_CONNTRACK: enabled (as module)
- CONFIG_NETFILTER_XT_MATCH_IPVS: enabled (as module)
- CONFIG_IP_NF_NAT: enabled (as module)
- CONFIG_NF_NAT: enabled (as module)
- CONFIG_NF_NAT_NEEDED: enabled
- CONFIG_POSIX_MQUEUE: enabled
- CONFIG_DEVPTS_MULTIPLE_INSTANCES: enabled

Optional Features:
- CONFIG_USER_NS: enabled
- CONFIG_SECCOMP: enabled
- CONFIG_CGROUP_PIDS: enabled
- CONFIG_MEMCG_SWAP: enabled
- CONFIG_MEMCG_SWAP_ENABLED: enabled
    (cgroup swap accounting is currently enabled)
- CONFIG_MEMCG_KMEM: enabled
- CONFIG_RESOURCE_COUNTERS: missing
- CONFIG_BLK_CGROUP: enabled
- CONFIG_BLK_DEV_THROTTLING: enabled
- CONFIG_IOSCHED_CFQ: enabled
- CONFIG_CFQ_GROUP_IOSCHED: enabled
- CONFIG_CGROUP_PERF: enabled
- CONFIG_CGROUP_HUGETLB: enabled
- CONFIG_NET_CLS_CGROUP: enabled
- CONFIG_NETPRIO_CGROUP: enabled
- CONFIG_CFS_BANDWIDTH: enabled
- CONFIG_FAIR_GROUP_SCHED: enabled
- CONFIG_RT_GROUP_SCHED: enabled
- CONFIG_IP_VS: enabled (as module)
- CONFIG_IP_VS_NFCT: enabled
- CONFIG_IP_VS_RR: enabled (as module)
- CONFIG_EXT3_FS: missing
- CONFIG_EXT3_FS_XATTR: missing
- CONFIG_EXT3_FS_POSIX_ACL: missing
- CONFIG_EXT3_FS_SECURITY: missing
    (enable these ext3 configs if you are using ext3 as backing filesystem)
- CONFIG_EXT4_FS: enabled (as module)
- CONFIG_EXT4_FS_POSIX_ACL: enabled
- CONFIG_EXT4_FS_SECURITY: enabled
- Network Drivers:
  - "overlay":
    - CONFIG_VXLAN: enabled (as module)
      Optional (for encrypted networks):
      - CONFIG_CRYPTO: enabled
      - CONFIG_CRYPTO_AEAD: enabled
      - CONFIG_CRYPTO_GCM: enabled (as module)
      - CONFIG_CRYPTO_SEQIV: enabled
      - CONFIG_CRYPTO_GHASH: enabled (as module)
      - CONFIG_XFRM: enabled
      - CONFIG_XFRM_USER: enabled
      - CONFIG_XFRM_ALGO: enabled
      - CONFIG_INET_ESP: enabled (as module)
      - CONFIG_INET_XFRM_MODE_TRANSPORT: enabled (as module)
  - "ipvlan":
    - CONFIG_IPVLAN: missing
  - "macvlan":
    - CONFIG_MACVLAN: enabled (as module)
    - CONFIG_DUMMY: enabled (as module)
  - "ftp,tftp client in container":
    - CONFIG_NF_NAT_FTP: enabled (as module)
    - CONFIG_NF_CONNTRACK_FTP: enabled (as module)
    - CONFIG_NF_NAT_TFTP: enabled (as module)
    - CONFIG_NF_CONNTRACK_TFTP: enabled (as module)
- Storage Drivers:
  - "aufs":
    - CONFIG_AUFS_FS: missing
  - "btrfs":
    - CONFIG_BTRFS_FS: enabled (as module)
    - CONFIG_BTRFS_FS_POSIX_ACL: enabled
  - "devicemapper":
    - CONFIG_BLK_DEV_DM: enabled (as module)
    - CONFIG_DM_THIN_PROVISIONING: enabled (as module)
  - "overlay":
    - CONFIG_OVERLAY_FS: enabled (as module)
  - "zfs":
    - /dev/zfs: missing
    - zfs command: missing
    - zpool command: missing

Limits:
- /proc/sys/kernel/keys/root_maxkeys: 1000000
---------------------------------------------------------------------------------------------------

**Comment by hehailong5** [comment]
docker info output:
----------------------------------------------------------------------------------------
-bash-4.2# docker info
Containers: 18
 Running: 17
 Paused: 0
 Stopped: 1
Images: 16
Server Version: 1.13.1
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: true
 Native Overlay Diff: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: active
 NodeID: vy1jrqptm2b44lzgxfd1wgbmv
 Is Manager: true
 ClusterID: wzcnfqwnymgucxm5l43mcuzfu
 Managers: 3
 Nodes: 3
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 120.0.0.183
 Manager Addresses:
  120.0.0.182:2377
  120.0.0.183:2377
  120.0.0.184:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: aa8187dbd3b7ad67d8e5e3a15115d3eef43a7ed1
runc version: 9df8b306d01f59d3a8029be411de015b7304dd8f
init version: 949e6fa
Kernel Version: 3.10.0-514.6.1.el7.x86_64
Operating System: NewStart Carrier Grade Server Linux
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.639 GiB
Name: controller
ID: 72BV:RCH5:XXAK:HTPS:QGNA:Z7AH:7Q5S:PTTU:R6XO:JJFV:XKCC:3QTO
Docker Root Dir: /root/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 10.67.18.252:33377
 127.0.0.0/8
Live Restore Enabled: false

**Comment by hehailong5** [comment]
B.T.W, I found these warnings in the dockerd log:
----------------------------------------------------------------------------------------------------
Mar  8 07:00:26 controller dockerd: time="2017-03-08T07:00:26.409299752+08:00" level=warning msg="Using pre-4.0.0 kernel for overlay2, mount failures may require kernel update"

Mar  8 07:00:27 controller dockerd: time="2017-03-08T07:00:27.097927013+08:00" level=warning msg="Not using native diff for overlay2: opaque flag erroneously copied up, consider update to kernel 4.8 or later to fix"

Mar  8 17:38:36 controller dockerd: time="2017-03-08T17:38:36.237311037+08:00" level=warning msg="memberlist: failed to receive: No installed keys could decrypt the message from=120.0.0.183:38816"
Mar  8 17:38:36 controller dockerd: time="2017-03-08T17:38:36.327066980+08:00" level=warning msg="memberlist: Decrypt packet failed: No installed keys could decrypt the message from=120.0.0.183:7946"
Mar  8 17:38:36 controller dockerd: time="2017-03-08T17:38:36.527189638+08:00" level=warning msg="memberlist: Decrypt packet failed: No installed keys could decrypt the message from=120.0.0.183:7946"

**Comment by thaJeztah** [comment]
The overlay2 messages are just warnings, so should not be an issue. The "decrypt" message is also a warning ("Decrypt packet failed: No installed keys could decrypt the message"), but @aboch can probably better tell if those are problematic.

**Comment by aboch** [comment]
The memberlist failures are about the networking control plane.
There was a bug because of which two nodes may end up with mismatched control plane encryption keys. @sanimej fixed that, he would be able to check whether you are hitting that.

**Comment by sanimej** [comment]
@hehailong5 In swarm mode there is an encrypted communication between all the nodes, including the managers. For whatever if there is are multiple rapid leader changes in the swarm cluster its possible for some worker nodes to go out of sync with the managers on what keys to use for encryption. This can cause `memberlist: failed to receive: No installed keys could decrypt the message` errors and also affect the connectivity for services that are created after this error starts happening. This has been fixed in Docker 17.03. Please upgrade the cluster to 17.03 CE and confirm that this error is not seen anymore and then try the connectivity between services.

**Comment by hehailong5** [comment]
Hi,

We use ansible to automate the following process:

1. set up swarm cluster
2. create custom overlay network
3. start the services attaching to the overlay network

Currently we didn't check the swarm cluster status and no sleep neither before step 2.
I'm wondering that "the swarm cluster is not ready" might make the overlay not work. is it possible? and how to ensure the cluster is ready to use overlay services?

**Comment by hehailong5** [comment]
Hi, same issue occurs this morning, when one service tries to connect to another service in the same overlay network via netcat, it complains "No route to Host". we have been using 1.12.3 and have no plan to upgrade it, we may want to know what commit(s) may probably fix this so we can merge them on top of 1.12.3.

**Comment by sam-thibault** [comment]
I don't see any recent activity on this issue.  I will close it as stale.
